### PR TITLE
Add interval to PercentilePruner.

### PR DESCRIPTION
**Pull request by VladSkripniuk**
Address #1136 .

## Description of the changes

1. Add `is_target_step` to `BasePruner` and implement `is_target_step` in `PercentilePruner`.
2. Use `pruner.is_target_step(epoch)` in `KerasPruningCallback` to skip pruning by interval.

### CODE REVIEW

File: optuna/pruners/percentile.py [hunk 1]
```diff
@@ -192,3 +192,10 @@ def prune(self, study, trial):
         if direction == StudyDirection.MAXIMIZE:
             return best_intermediate_result < p
         return best_intermediate_result > p
+
+    def is_target_step(self, step):
+        # type: (int) -> bool
+        if step < self._n_warmup_steps:
+            return False
+
+        return (step - self._n_warmup_steps) % self._interval_steps == 0
```
Comment (toshihikoyanase) [inline optuna/pruners/percentile.py:201]: Please add a test case to confirm if the target steps determined by this function matches with those determined by the `prune` method? I think the test cases can be written using the [`DeteministicPruner(True)`](https://github.com/optuna/optuna/blob/master/optuna/testing/integration.py#L4-L13).

IMO, the current logic does not match with `prune` in some cases. For example, the check corresponding to L198 seems to be `if step <= self._nwarmup_steps` in the `prune` method.

**Comment by toshihikoyanase** [comment]
@keisuke-umezawa Could you give us your feedback about https://github.com/optuna/optuna/pull/1234#pullrequestreview-411385804? I think it is controversial.

> I think we need to ensure that the target steps determined by `is_target_step` are actual targets of the `prune` method.

**Comment by keisuke-umezawa** [comment]
@toshihikoyanase @VladSkripniuk 

> I think we need to ensure that the target steps determined by is_target_step are actual targets of the prune method.

I agree with the idea. Somehow, we should guarantee it.

One idea is:

```
    def is_target_step(self):
        # type: (int) -> bool

        all_trials = study.get_trials(deepcopy=False)
        n_trials = len([t for t in all_trials if t.state == TrialState.COMPLETE])

        if n_trials == 0:
            return False

        if n_trials < self._n_startup_trials:
            return False

        step = trial.last_step
        if step is None:
            return False

        n_warmup_steps = self._n_warmup_steps
        if step <= n_warmup_steps:
            return False

        if not _is_first_in_interval_step(
            step, trial.intermediate_values.keys(), n_warmup_steps, self._interval_steps
        ):
            return False

        return (step - self._n_warmup_steps) % self._interval_steps == 0

    def prune(self, study, trial):
        # type: (Study, FrozenTrial) -> bool
        
        if not is_target_step(epoch):
            return False

        direction = study.direction
        best_intermediate_result = _get_best_intermediate_result_over_steps(trial, direction)
        ...
```

**Comment by toshihikoyanase** [comment]
@keisuke-umezawa 

Thank you for your proposal. It seems to be simple and straightforward, but I'm a bit concerned about a performance issue.

If `is_target_step` invokes `study.get_trials`, it takes communication time to access the storage. Besides, the number of `study.get_trials` calls will be doubled (i.e., in `prune` and `is_target_step`) in the worst case. So, I'd like to detect the target steps without accessing the storage if possible. Does it make sense?

**Comment by keisuke-umezawa** [comment]
> So, I'd like to detect the target steps without accessing the storage if possible. Does it make sense?

I totally agree with it.

**Comment by toshihikoyanase** [comment]
@keisuke-umezawa Thank you for your proposal! It sounds good to me.

**Comment by toshihikoyanase** [comment]
@VladSkripniuk What do you think of @keisuke-umezawa 's suggestion? I think it's viable, but do you have any concerns?

**Comment by toshihikoyanase** [comment]
@keisuke-umezawa Are you interested in a new PR with your proposed code?
This PR has been stale for a while and I'm thinking of the restart of the code review with it.

**Comment by Crissman** [comment]
@VladSkripniuk This PR is stale and has conflicts with master, so closing for now. If there's further interest, we may want to take a look at #1161 (comment)

Please re-open if you want to proceed and resolve the conflicts.

**Pull request closed without merging**